frontend/src/metabase/dashboard/components/EmbedMenu/EmbedMenu.tsx

```diff
@@ -0,0 +1,151 @@
+  const isPublicSharingEnabled = useSelector(state =>
+    getSetting(state, "enable-public-sharing"),
+  );
+
+  const isEmbeddingEnabled = useSelector(state =>
+    getSetting(state, "enable-embedding"),
+  );
+
+  const isAdmin = useSelector(getUserIsAdmin);
+
+  const shouldRenderForUser = isAdmin || props.hasPublicLink;
+
+  if (
+    isEmbeddingEnabled == null ||
+    isPublicSharingEnabled == null ||
+    !shouldRenderForUser
+  ) {
+    return null;
+  }
+
+  return <EmbedMenuInner {...props} />;
+};
+
+const EmbedMenuInner = ({ hasPublicLink, onModalOpen }: EmbedMenuProps) => {
+  const initialMenuMode: EmbedMenuModes = useEmbedMenuMode({
+    hasPublicLink,
+  });
+
+  const [isOpen, setIsOpen] = useState(false);
+  const [menuMode, setMenuMode] = useState(initialMenuMode);
+
+  const onMenuSelect = (menuMode: EmbedMenuModes) => {
+    setIsOpen(true);
+    setMenuMode(menuMode);
+  };
+
+  const renderEmbedMenu = () => (
+    <DashboardEmbedHeaderMenu
+      hasPublicLink={hasPublicLink}
+      openPublicLinkPopover={() => onMenuSelect("public-link-popover")}
+      openEmbedModal={onModalOpen}
+      target={<DashboardEmbedHeaderButton />}
+    />
+  );
+
+  const renderEmbedModalTrigger = () => (
+    <DashboardEmbedHeaderButton onClick={onModalOpen} />
+  );
+
+  const onClosePublicLinkPopover = () => {
+    setIsOpen(false);
+    setMenuMode(initialMenuMode);
+  };
+
+  const renderPublicLinkPopover = () => {
+    return (
+      <Popover opened={isOpen} onClose={onClosePublicLinkPopover}>
+        <Popover.Target>
+          <DashboardEmbedHeaderButton
+            onClick={isOpen ? onClosePublicLinkPopover : () => setIsOpen(true)}
+          />
+        </Popover.Target>
+        <Popover.Dropdown>
+          <div>Public Link Popover</div>
+        </Popover.Dropdown>
+      </Popover>
+    );
+  };
+
+  const renderEmbeddingDisabled = () => {
+    return <DashboardEmbedHeaderButton disabled />;
+  };
+
+  const getEmbedContent = (menuMode: EmbedMenuModes) => {
+    if (menuMode === "embed-menu") {
+      return renderEmbedMenu();
+    } else if (menuMode === "embed-modal") {
+      return renderEmbedModalTrigger();
+    } else if (menuMode === "public-link-popover") {
+      return renderPublicLinkPopover();
+    } else if (menuMode === "embedding-disabled") {
+      return renderEmbeddingDisabled();
+    }
+
+    return null;
+  };
+
+  return getEmbedContent(menuMode);
+};
+
+const useEmbedMenuMode = ({
+  hasPublicLink,
+}: {
+  hasPublicLink: boolean;
+}): EmbedMenuModes => {
+  const isPublicSharingEnabled = useSelector(state =>
+    getSetting(state, "enable-public-sharing"),
+  );
+
+  const isEmbeddingEnabled = useSelector(state =>
+    getSetting(state, "enable-embedding"),
+  );
+
+  const isAdmin = useSelector(getUserIsAdmin);
+
+  if (isAdmin) {
+    if (!isEmbeddingEnabled) {
+      return "embedding-disabled";
+    }
+    return isPublicSharingEnabled ? "embed-menu" : "embed-modal";
```

Comment:
Maybe instead of mixing if and ternary operators, we only use ifs.

```suggestion
    if (!isEmbeddingEnabled) {
      return "embedding-disabled";
    }
    
    if (isPublicSharingEnabled) {
      return "embed-menu"
    }
    
    return "embed-modal";
```

Previously it was quite acceptable to use ternary operator because there was only one boolean flag, but now we have 2.

What do you think?

Reply:
Yep looks good 😄